File: src/Lexer.cpp
#include <gazebo.hpp>
#include <gazebo/Lexer.hpp>
#include <vector>

namespace gazebo
{

void Lexer::line()
{
    while (next() != '\n')
        ;
}

Token Lexer::identifier()
{
    return token(Token::Type::IDENTIFIER);
}

Token Lexer::number()
{
    bool isfloat = false;

    while (isdigit(peek())) {
        next();

        if (!isfloat && match('.')) {
            isfloat = true;
            G_ASSERT(isdigit(peek()));
        }
    }

    return token(Token::Type::NUMBER);
}

Token Lexer::string()
{
    return token(Token::Type::STRING);
}

Token Lexer::lex()
{
    char ch = next();

    if (!ch)
        return token(Token::Type::END);

    if (isspace(ch)) {
        while (isspace(peek()))
            next();

        return token(Token::Type::WHITESPACE);
    }

    if (ch == '#') {
        line();
        return token(Token::Type::COMMENT);
    }

    switch (ch) {
    case '(':
        return token(Token::Type::PAREN_OPEN);

    case ')':
        return token(Token::Type::PAREN_CLOSE);

    case '[':
        return token(Token::Type::BRACKET_OPEN);

    case ']':
        return token(Token::Type::BRACKET_CLOSE);

    case '{':
        return token(Token::Type::BRACE_OPEN);

    case '}':
        return token(Token::Type::BRACE_CLOSE);

    case ':':
        return token(Token::Type::COLON);

    case ';':
        return token(Token::Type::SEMICOLON);

    case '.':
        return token(Token::Type::DOT);

    case ',':
        return token(Token::Type::COMMA);

    case '=':
        return match('=') ? token(Token::Type::EQUAL_EQUAL) : token(Token::Type::EQUAL);

    case '!':
        return match('=') ? token(Token::Type::BANG_EQUAL) : token(Token::Type::BANG);

    case '<':
        return match('=') ? token(Token::Type::LESS_EQUAL) : token(Token::Type::LESS);

    case '>':
        return match('=') ? token(Token::Type::GREATER_EQUAL) : token(Token::Type::GREATER);

    case '+':
        return token(Token::Type::PLUS);

    case '-':
        return token(Token::Type::MINUS);

    case '*':
        return token(Token::Type::STAR);

    case '/':
        if (check('/')) {
            line();
            return token(Token::Type::COMMENT);
        }

        return token(Token::Type::SLASH);
    }

    if (isdigit(ch))
        return number();

    G_UNREACHED();
}

std::vector<Token> Lexer::lex_all()
{
    std::vector<Token> tokens;

    Token token;

    do {
        token = lex();

        if (!token.is(Token::Type::WHITESPACE))
            tokens.push_back(token);

    } while (!token.is(Token::Type::END));

    return tokens;
}

}  // namespace gazebo

File: parser/lexer.go
package parser

import (
	"errors"
	"fmt"
	"gazebo/ast"
	"regexp"
	"unicode"
)

type lexer struct {
	source   string
	position struct {
		last    ast.Position
		current ast.Position
	}
	program *ast.Program
	err     error
}

var _ yyLexer = (*lexer)(nil)

func (l *lexer) Lex(lval *yySymType) int {
	const eof = 0

	for len(l.source) > 0 && unicode.IsSpace(rune(l.source[0])) {
		if l.source[0] == '\n' {
			l.position.current.Line++
			l.position.current.Column = 0
		}

		l.position.current.Column++
		l.position.current.Offset++

		l.source = l.source[1:]
	}

	lval.Position = l.position.current
	l.position.last = l.position.current

	for _, rule := range rules {
		if rule.regexp.MatchString(l.source) {
			lval.Lexeme = rule.regexp.FindString(l.source)

			l.position.current.Column += len(lval.Lexeme)
			l.position.current.Offset += len(lval.Lexeme)

			l.source = l.source[len(lval.Lexeme):]

			return rule.token
		}
	}

	return eof
}

func (l *lexer) Error(s string) {
	l.err = errors.Join(fmt.Errorf("parser: %s: %s", l.position.last, s))
}

func exp(s string) *regexp.Regexp { return regexp.MustCompile("^" + s) }

func lit(s string) *regexp.Regexp { return exp(regexp.QuoteMeta(s)) }

var rules = []struct {
	token  int
	regexp *regexp.Regexp
}{
	{SEMICOLON, lit(`;`)},
	{COMMA, lit(`,`)},
	{COMMENT, exp(`//.*`)},
	{BRACE_OPEN, lit(`{`)},
	{BRACE_CLOSE, lit(`}`)},
	{PAREN_OPEN, lit(`(`)},
	{PAREN_CLOSE, lit(`)`)},
	{BRACKET_OPEN, lit(`[`)},
	{BRACKET_CLOSE, lit(`]`)},
	{PLUS, lit(`+`)},
	{MINUS, lit(`-`)},
	{STAR, lit(`*`)},
	{SLASH, lit(`/`)},
	{PERCENT, lit(`%`)},
	{EQUAL_EQUAL, lit(`==`)},
	{EQUAL, lit(`=`)},
	{BANG_EQUAL, lit(`!=`)},
	{BANG, lit(`!`)},
	{LESS_EQUAL, lit(`<=`)},
	{LESS, lit(`<`)},
	{GREATER_EQUAL, lit(`>=`)},
	{GREATER, lit(`>`)},
	{NULL, lit(`null`)},
	{FALSE, lit(`false`)},
	{TRUE, lit(`true`)},
	{AND, lit(`and`)},
	{IF, lit(`if`)},
	{ELSE, lit(`else`)},
	{WHILE, lit(`while`)},
	{OR, lit(`or`)},
	{RETURN, lit(`return`)},
	{FLOAT, exp(`[0-9]*\.[0-9]+`)},
	{INTEGER, exp(`[0-9]+`)},
	{STRING, exp(`("(?:\\.|[^"])*"|'(?:\\.|[^'])*')`)},
	{IDENTIFIER, exp(`[a-zA-Z_][a-zA-Z0-9_]*`)},
}
